cpp/src/parquet/column_chunker.h

```diff
@@ -0,0 +1,168 @@
+/// same blob is encountered again, the system can refer to the hash instead of physically
+/// storing the blob again. In the example above, the CAS system would phiysically store
+/// Page1, Page2, Page3, and Page4 only once and the required metadata to reassemble the
+/// files.
+/// While the deduplication is performed by the CAS system, the parquet chunker makes it
+/// possible to efficiently deduplicate the data by consistently dividing the data into
+/// chunks.
+///
+/// Implementation details:
+///
+/// Only the parquet writer must be aware of the content defined chunking, the reader
+/// doesn't need to know about it. Each parquet column writer holds a
+/// ContentDefinedChunker instance depending on the writer's properties. The chunker's
+/// state is maintained across the entire column without being reset between pages and row
+/// groups.
+///
+/// The chunker receives the record shredded column data (def_levels, rep_levels, values)
+/// and goes over the (def_level, rep_level, value) triplets one by one while adjusting
+/// the column-global rolling hash based on the triplet. Whenever the rolling hash matches
+/// a predefined mask, the chunker creates a new chunk. The chunker returns a vector of
+/// Chunk objects that represent the boundaries of the chunks///
+/// Note that the boundaries are deterministically calculated exclusively based on the
+/// data itself, so the same data will always produce the same chunks - given the same
+/// chunker configuration.
+///
+/// References:
+/// - FastCDC paper: "FastCDC: a Fast and Efficient Content-Defined Chunking Approach for
+/// Data Deduplication"
+///   https://www.usenix.org/system/files/conference/atc16/atc16-paper-xia.pdf
+class ContentDefinedChunker {
+ public:
+  /// Create a new ContentDefinedChunker instance
+  ///
+  /// @param level_info Information about definition and repetition levels
+  /// @param size_range Min/max chunk size as pair<min_size, max_size>, the chunker will
+  ///                   attempt to uniformly distribute the chunks between these extremes.
+  /// @param norm_factor Normalization factor to center the chunk size around the average
+  ///                    size more aggressively. By increasing the normalization factor,
+  ///                    probability of finding a chunk boundary increases.
+  ContentDefinedChunker(const LevelInfo& level_info,
+                        std::pair<uint64_t, uint64_t> size_range,
+                        uint8_t norm_factor = 0);
+
+  /// Get the chunk boundaries for the given column data
+  ///
+  /// @param def_levels Definition levels
+  /// @param rep_levels Repetition levels
+  /// @param num_levels Number of levels
+  /// @param values Column values as an Arrow array
+  /// @return Vector of Chunk objects representing the chunk boundaries
+  const ::arrow::Result<std::vector<Chunk>> GetBoundaries(const int16_t* def_levels,
```

Comment:
When will the `::arrow::Result` hold an error? Usually in the parquet module we use throw instead of status.

Reply:
When the input array's type is unsupported, like nested arrays. Updating to raise instead.

Reply:
Updated to throw a parquet exception instead.